components/mpas-ocean/src/shared/mpas_ocn_time_average_coupled.F

```diff
@@ -473,7 +490,52 @@ subroutine ocn_time_average_coupled_accumulate(statePool, forcingPool, timeLevel
            end do
            !$omp end do
            !$omp end parallel
-        endif
+        elseif (trim(config_glc_thermal_forcing_coupling_mode) == '3d') then
+           call mpas_pool_get_array(forcingPool, 'avgThermalForcingAtZLevels', avgThermalForcingAtZLevels)
+           call mpas_pool_get_array(forcingPool, 'avgThermalForcingAtZLevelsMask', avgThermalForcingAtZLevelsMask)
+           call mpas_pool_get_array(forcingPool, 'glcZLevels', glcZLevels)
+           call mpas_pool_get_subpool(statePool, 'tracers', tracersPool)
+           call mpas_pool_get_array(tracersPool, 'activeTracers', activeTracers, 2)
+           call mpas_pool_get_dimension(tracersPool, 'index_temperature', indexTemperature)
+           call mpas_pool_get_dimension(tracersPool, 'index_salinity', indexSalinity)
+           call mpas_pool_get_dimension(tracersPool, 'nGlcZLevels', nGlcZLevels)
+
+           !$omp parallel
+           !$omp do schedule(runtime)
+           do iLevelGlc = 1, nGlcZLevels
+              ! find vertical level that is just above the critical depth reference level
+              ! this does not account for depression due to ice shelf cavities or sea ice
+              iLevelCritDepth = nVertLevels  ! default to deepest layer if we don't find the desired depth
+              do iLevel = 1, nVertLevels
+                 if(refBottomDepth(iLevel) > glcZLevels(iLevelGlc)) then
+                    iLevelCritDepth = iLevel
+                    exit
+                 end if
+              end do
```

Comment:
This determines the level using `refBottomDepth`, which is only appropriate for a pure z-level coordinate.  Anywhere near Antarctic ice shelves, this will be badly wrong.  In cavities, it will be much worse.

Instead, the right level needs to be determined for each cell independently, based on which zMid is closest to the given `glcZLevels`.  I can help with this but this needs to be fixed before this PR can move over to E3SM.

Reply:
Rather than finding the nearest depth, it might be worth linearly interpolating to `glcZLevels`, but that could wait for a follow-up PR if you prefer.